lukasz-2196-finished-test-coverage-and-feature fixes #2196

lukaszgryglicki opened the pull request:
**This implements: https://github.com/kubernetes/test-infra/issues/2196**:

fixes #2196

**Special notes for your reviewer**:
Details: https://github.com/kubernetes/test-infra/issues/2196

**Release note**:
No release notes needed imho.

lukaszgryglicki commented:
I've signed the CLA just now.
Also this PR doesn't need Release Notes.
It is just an improvement to one of prow bot's plugins (`label`).
Not sure what to do now.

spxtr commented:
@k8s-bot ok to test

grodrigues3 commented on commands.md:
```diff
@@ -7,8 +7,12 @@ Command | Implemented By | Who can run it | Description
 `/assign [@userA @userB @etc]` | prow [assign](./prow/plugins/assign) | kubernetes org members | Assigns specified people (or yourself if no one is specified)
 `/unassign [@userA @userB @etc]` | prow [assign](./prow/plugins/assign) | kubernetes org members | Unassigns specified people (or yourself if no one is specified)
 `/area` | prow [label](./prow/plugins/label) | anyone | adds an area/<> label if it exists
+`/remove-area` | prow [label](./prow/plugins/label) | anyone | removes an area/<> label if it exists
```
What if there are multiple area labels?  Does this just remove all of them?

grodrigues3 replied:
I read the code and I understand.  Make it clear in the docs that these commands take arg(s) that are the label suffixes.

grodrigues3 commented:
@foxish any idea why the CLA didn't work?

foxish commented:
Retrying CLA...

foxish commented:
This one is at the CNCF bot end. I'd recommend emailing helpdesk@rt.linuxfoundation.org

lukaszgryglicki commented:
I've already asked CNCF helpdesk about 6 hours ago.
Hopefully they'll help.
I think I just need an account like lukaszgryglicki@cncf.org or lukaszgryglicki@linuxfoundation.orb....
But really I prefer to stay with my old lukaszgryglicki@o2.pl approved by CNCF.

lukaszgryglicki commented:
Updated docs as requested.

lukaszgryglicki commented:
/release-note-none

thelinuxfoundation commented:
bump the CLA check

lukaszgryglicki commented:
@thelinuxfoundation thanks.
BTW: how can I join kubernetes org?

cblecker commented:
I'm not sure removing `do-not-merge` via a command like this is safe. There are lots of reasons maintainers apply this label, and it is likely something that should be left for maintainers to remove.

lukaszgryglicki commented:
Well that was requested by @grodrigues3 no problem to remove that functionality.

spxtr commented:
We need to double-check that the munger won't merge if `release-note-needed` is on the PR. Then we can just make the release-note munger not apply `do-not-merge`.

cblecker commented:
I agree, @spxtr.. Specific on the release note issue: #2368

spxtr commented:
@lukaszgryglicki do you want to take care of that? It can happen in a separate PR or in this one, I don't mind either way.

grodrigues3 commented:
On second thought, it may be safer to err on the side of caution and not allow removal of do-not-merge via command.

lukaszgryglicki commented:
So I will remove functionality `/remove-do-not-merge` and if I understand correctly:
- make sure that munger will not merge if there is `release-note-needed` Label
- release-note munger will not apply `do-not-merge` (unconditionally? just remove that functionality)? 

I prefer 2 separate PRs. I prefer smaller PRs with just single functionality so:
- I'll remove `/remove-do-not-merge` from this PR
- I'll open another one for #2368 

Will do tomorrow.

cblecker commented:
Yes, that's my understanding of the direction we want to go in.

lukaszgryglicki commented:
So I've removed support for `/remove-do-not-merge` from this PR.
It is ready for re-review.
I'll implement rest in a separate PR for #2368

grodrigues3 commented:
/lgtm

spxtr commented:
WRT mac builds, this might be fixed in a newer version of `rules_go`. If not, you should be able to test with go with `make test`.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -107,7 +114,30 @@ func handle(gc githubClient, log *logrus.Entry, ic github.IssueCommentEvent) err
 		}
+	}
+
+	labelsToRemove := []string{}
```
Prefer `var labelsToRemove []string`

lukaszgryglicki replied:
Will fix (tomorrow) gotta go now

lukaszgryglicki commented:
Rebased.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -107,7 +114,30 @@ func handle(gc githubClient, log *logrus.Entry, ic github.IssueCommentEvent) err
+
+	labelsToRemove := []string{}
+	for _, match := range removeLabelMatches {
+		for _, removeLabel := range strings.Split(match[0], " ")[1:] {
```
it would be really nice if we could reuse this stuff, since right now the code "select the labels to be added" and "select the labels to be removed" is duplicated.

lukaszgryglicki replied:
Will take a look tomorrow to extract this into some func/method

lukaszgryglicki commented:
@spxtr addressed Your feedback, extracted common code (RE parsing to get labels) to new function `getLabelsFromREMatches` and refactored adding/removing labels to use this.
Code to iterate over labels to add / remove remained in `handle` function because logic is different for add and for remove and I wanted to avoid creating another function for that with big `IF` switch (only changed to use values returned from `getLabelsFromREMatches`).

spxtr commented:
Thanks. This will be deployed when we next bump the hook version, maybe after #2474 goes in?

lukaszgryglicki commented:
OK thanks.

grodrigues3 commented:
@lukaszgryglicki thanks for getting this done!

MrHohn commented:
Thanks for the great feature!
